File: exercises/exercises-mpi-p/pingpong.py
#!/usr/bin/env python
################################################################
################################################################
####
#### This program file is part of the book and course
####   Parallel Computing for Science and Engineering
#### by Victor Eijkhout, copyright 2013-2022
####
#### pingpong.py : MPI python exercise for send/recv
####
################################################################
################################################################

from mpi4py import MPI
import numpy as np
import time

comm = MPI.COMM_WORLD

nprocs = comm.Get_size()
procno = comm.Get_rank()
ntests = 10000

if procno==0 or procno==nprocs-1:
    print("Node name:",MPI.Get_processor_name())

## Determine who will be communicating:
## set processA, processB
##solution
processA = 0
processB = nprocs-1
##solved

##
## Do a range of test sizes
##

for s in [1,10,100,1000,10000,100000,1000000]:
    if procno==processA:
        ####
        #### Originating process:
        ####
        ##
        ## create send_data and recv_data numpy arrays
        ##
        ##solution
        send_data = np.arange(s, dtype=np.float64)
        recv_data = np.empty(s, dtype=np.float64)
        ##solved
        ## initialize send data buffer
        for i in range(s):
            send_data[i] = i+1
        starttime = MPI.Wtime()
        for test in range(ntests):
            ##
            ## Originating process:
            ## send the `ping' and receive the `pong'
            comm.Send([send_data,MPI.DOUBLE],dest=
                      ##solution
                      processB
                      ##solved
            )
            comm.Recv([recv_data,MPI.DOUBLE],source=
                      ##solution
                      processB
                      ##solved
            )
        elapsed = MPI.Wtime()-starttime
        ##
        ## Report time and detect errors
        ##
        print("Size=%d, elapsed time: %e" % (s,elapsed))
        c = send_data==recv_data
        if not c.all():
            print("oops",send_data,recv_data)
    elif procno==processB:
        ####
        #### Other process
        ####
        ##
        ## create thru_data array
        ##
        ##solution
        thru_data = np.empty(s, dtype=np.float64)
        ##solved
        for test in range(ntests):
            ##
            ## receive the `ping' and send `pong'
            ##
            comm.Recv([thru_data,MPI.DOUBLE],source=
                      ##solution
                      processA
                      ##solved
            )
            comm.Send([thru_data,MPI.DOUBLE],dest=
                      ##solution
                      processA
                      ##solved
            )

File: exercises/exercises-mpl-mpl/pingpong.cxx
/****************************************************************
 ****
 **** This program file is part of the book 
 **** `Parallel programming for Science and Engineering'
 **** by Victor Eijkhout, eijkhout@tacc.utexas.edu
 ****
 **** copyright Victor Eijkhout 2012-2026
 ****
 **** MPI Exercise using MPL layouts on a pingpong
 ****
 ****************************************************************/

#include <iostream>
#include <iomanip>
#include <mpi.h>
using namespace std;
#include <print>
#include <vector>
#include <mpl/mpl.hpp>

int main() {

  const mpl::communicator &comm_world = mpl::environment::comm_world();
  int nprocs, procno;
  nprocs = comm_world.size();
  procno = comm_world.rank();

#define NEXPERIMENTS 100
#ifndef SIZE
#define SIZE 10000
#endif

  // Exercise:
  // -- define a layout for the buffer
  // -- use the layout to send the buffer
  int processA,processB;
  processA = 0; processB = nprocs-1;
  double t;
  std::vector<double> send(SIZE,1.1), recv(SIZE,1.1);

  // Define an mpl::contiguous_layout for send/recv operations
/**** your code here ****/

  if (procno==processA) {
    t = MPI_Wtime();
    for (int n=0; n<NEXPERIMENTS; n++) {
      comm_world.send
/**** your code here ****/
      comm_world.recv
/**** your code here ****/
    }
    t = MPI_Wtime()-t; t /= NEXPERIMENTS;
    {
      int nanosec = t*1000000000;
      std::println
	( "Time for pingpong of size {}: {} (usec)",
	  SIZE,nanosec*1.e-3 );
	  // fixed << setprecision(3) << 
    }
  } else if (procno==processB) {
    for (int n=0; n<NEXPERIMENTS; n++) {
      comm_world.recv
/**** your code here ****/
      comm_world.send
/**** your code here ****/
    }
  }

  return 0;
}
